File: backtester/crates/bt-core/src/reason_codes.rs
//! Canonical, enumerated reason codes for trade log records.
//!
//! These codes are intended to be stable across releases and suitable for downstream analytics
//! (JSON reports, CSV exports, dashboards). They intentionally do not encode strategy-specific
//! free-form text.

use serde::Serialize;

/// Canonical reason code for a trade log record.
#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Serialize)]
#[serde(rename_all = "snake_case")]
pub enum ReasonCode {
    // Entries / position changes
    EntrySignal,
    EntrySignalSubBar,
    EntryPyramid,

    // Exits
    ExitStopLoss,
    ExitTakeProfit,
    ExitTrailingStop,
    ExitSignalFlip,
    ExitFilter,
    ExitFunding,
    ExitForceClose,
    ExitEndOfBacktest,

    // Non-trade balance events
    FundingPayment,

    Unknown,
}

/// Classify a (action, reason) pair into a stable [`ReasonCode`].
pub fn classify_reason_code(action: &str, reason: &str) -> ReasonCode {
    if action == "FUNDING" {
        return ReasonCode::FundingPayment;
    }

    // Entries
    if action.starts_with("OPEN_") {
        if reason.contains("sub-bar") {
            return ReasonCode::EntrySignalSubBar;
        }
        return ReasonCode::EntrySignal;
    }
    if action.starts_with("ADD_") {
        return ReasonCode::EntryPyramid;
    }

    // Exits (includes partial exits via REDUCE_*)
    if action.starts_with("CLOSE_") || action.starts_with("REDUCE_") {
        if reason.contains("Stop Loss") {
            return ReasonCode::ExitStopLoss;
        }
        if reason.contains("Trailing Stop") {
            return ReasonCode::ExitTrailingStop;
        }
        if reason.contains("Take Profit") {
            return ReasonCode::ExitTakeProfit;
        }
        if reason.contains("Signal Flip") {
            return ReasonCode::ExitSignalFlip;
        }
        if reason.contains("Funding") {
            return ReasonCode::ExitFunding;
        }
        if reason.contains("Force Close") {
            return ReasonCode::ExitForceClose;
        }
        if reason.contains("End of Backtest") {
            return ReasonCode::ExitEndOfBacktest;
        }
        return ReasonCode::ExitFilter;
    }

    ReasonCode::Unknown
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn classify_entries() {
        assert_eq!(
            classify_reason_code("OPEN_LONG", "High entry"),
            ReasonCode::EntrySignal
        );
        assert_eq!(
            classify_reason_code("OPEN_SHORT", "High entry (sub-bar)"),
            ReasonCode::EntrySignalSubBar
        );
        assert_eq!(
            classify_reason_code("ADD_LONG", "Pyramid #1"),
            ReasonCode::EntryPyramid
        );
    }

    #[test]
    fn classify_exits() {
        assert_eq!(
            classify_reason_code("CLOSE_LONG", "Stop Loss"),
            ReasonCode::ExitStopLoss
        );
        assert_eq!(
            classify_reason_code("CLOSE_LONG", "Trailing Stop"),
            ReasonCode::ExitTrailingStop
        );
        assert_eq!(
            classify_reason_code("REDUCE_LONG", "Take Profit (Partial)"),
            ReasonCode::ExitTakeProfit
        );
        assert_eq!(
            classify_reason_code("CLOSE_SHORT", "Signal Flip"),
            ReasonCode::ExitSignalFlip
        );
        assert_eq!(
            classify_reason_code("CLOSE_SHORT", "Funding Headwind"),
            ReasonCode::ExitFunding
        );
        assert_eq!(
            classify_reason_code("CLOSE_SHORT", "End of Backtest"),
            ReasonCode::ExitEndOfBacktest
        );
        assert_eq!(
            classify_reason_code("CLOSE_SHORT", "Trend Breakdown"),
            ReasonCode::ExitFilter
        );
    }

    #[test]
    fn classify_funding_payment() {
        assert_eq!(
            classify_reason_code("FUNDING", "Funding rate=0.0001"),
            ReasonCode::FundingPayment
        );
    }
}

File: tools/reason_codes.py
#!/usr/bin/env python3
"""Canonical reason-code classifier shared by replay audit tooling.

This module mirrors the canonical reason mapping used by the Rust backtester
(`bt-core/src/reason_codes.rs`) so that live/paper/backtester audit reports
use one stable, machine-readable reason vocabulary.
"""

from __future__ import annotations


def classify_reason_code(action_code: str, reason: str) -> str:
    """Classify an `(action_code, reason)` pair into a canonical reason code."""
    action = str(action_code or "").strip().upper()
    reason_text = str(reason or "")

    if action == "FUNDING":
        return "funding_payment"
    if action.startswith("OPEN_"):
        if "sub-bar" in reason_text:
            return "entry_signal_sub_bar"
        return "entry_signal"
    if action.startswith("ADD_"):
        return "entry_pyramid"
    if action.startswith("CLOSE_") or action.startswith("REDUCE_"):
        if "Stop Loss" in reason_text:
            return "exit_stop_loss"
        if "Trailing Stop" in reason_text:
            return "exit_trailing_stop"
        if "Take Profit" in reason_text:
            return "exit_take_profit"
        if "Signal Flip" in reason_text:
            return "exit_signal_flip"
        if "Funding" in reason_text:
            return "exit_funding"
        if "Force Close" in reason_text:
            return "exit_force_close"
        if "End of Backtest" in reason_text:
            return "exit_end_of_backtest"
        return "exit_filter"
    return "unknown"
